### ISSUE TITLE
Group highlighting highlights wrong portions of the text

### ISSUE THREAD

**Issue by AbraxasDMG** [dnGrep/dnGrep#1069]
Hi everyone.

I've been playing a bit with regular expressions and tested some more complex examples (more complex for me, at least) and found that the group highlighting doesn't seem to accurately work.

I used the following regex pattern with _Case sensitive_, _Multiline_ and _Highlighting Groups_ modifiers

`^(?!.*#.*)\s*auth\s*required\s*pam_faillock\.so.*deny=([-+]?\d+)\s+unlock_time=([-+]?\d+).*$`

against the following text
```
auth        required                                     pam_deny.so
auth        required      pam_faillock.so preauth silent audit deny=11 unlock_time=-3620
auth        required      pam_faillock.so preauth silent audit deny=7 unlock_time=320
account     required                                     pam_unix.so    broken_shadow
```

and the result looks like this

![image](https://github.com/dnGrep/dnGrep/assets/8733868/66a6d1e7-57de-476c-8ea0-7f8d238a779b)

when it actually should look like this

![image](https://github.com/dnGrep/dnGrep/assets/8733868/7d4ba604-a5bc-4b15-97b0-02a771192f46)

If I add additional text into the Simple input text field, the highlighting moves

![image](https://github.com/dnGrep/dnGrep/assets/8733868/4da5818f-7118-46e1-9c05-2d6e9124fff9)

If I uncheck _Multiline_, the highlighting works properly

![image](https://github.com/dnGrep/dnGrep/assets/8733868/2f750022-118c-45db-a7e2-525bebca24b5)

Can this be fixed?
Thanks!

**Comment by doug24** [comment]
Thanks for the report and data to reproduce it.  I agree it should be fixed and will look into it.  But I wonder why you have multiline checked when your regex is anchored to the start and end of a line?  You don't need to have multiline.

It appears to be a problem with the EOL characters - regex uses just `\n` and Windows newline is `\r\n`

**Comment by AbraxasDMG** [comment]
Hi. Thank you very much for fast reply.

The 'm' flag stands for "multiline" and is used to modify the behavior of the start-of-line (^) and end-of-line ($) anchors in a regular expression. By default, these anchors match the start and end of the input string only. However, when the 'm' flag is used, the ^ and $ anchors also match the start and end of each line within the input string.

In combination with 'g' flag, the regex pattern is matches against each line within multiline text.

**Comment by fade2gray** [comment]
I think you are misunderstanding the usage of the `'g'` flag.

> 'g': Global search (finds all matches in the input string) ([reference ](https://codedamn.com/news/javascript/what-does-m-flag-mean-in-regular-expressions-regex))

... regardless whether the `'m'` flag is used.

![Screenshot 2024-03-06 234528](https://github.com/dnGrep/dnGrep/assets/1332009/f4f2da96-7601-4806-ac73-3ce00392274e)![Screenshot 2024-03-06 234634](https://github.com/dnGrep/dnGrep/assets/1332009/54a8962b-19d8-4400-998c-dbab9be21565)

https://regex101.com/r/nNjtfv/1

**Comment by doug24** [comment]
dnGrep uses .NET regular expressions, and there is no `g` flag - but the default behavior is the same as having set the `g` flag:
<img src="https://github.com/dnGrep/dnGrep/assets/17227248/1b603c82-a1d8-470e-bcd6-bad8977a0cea" width="460"/>

The use of multiline in dnGrep may be different than you are thinking - and maybe not standard - but this is what it does:  
- Without Multiline checked, dnGrep reads the file (or sample input from the test dialog) line by line and applies the regex to each line. 
- With Multiline checked, dnGrep reads the whole file and applies the regex once to all the lines.  Since .NET regex is global, it will find all the matches on all the lines.  This is necessary if your need to search across lines.  

With the ^ and $ anchors, you should get the same results in dnGrep with or without multiline checked.

I think the ugliest part of .NET regex is that `$` matches the *nix newline `\n` but does not match the Windows newline `\r\n`.  I'm sure they had a good reason to do that, but it's the source of endless bugs or unexpected behavior. There is code in dnGrep to try to fix this for the user - it converts the file to `\n` before searching (keeping track of what lines were changed) and then converts back to the original newlines and bumping the positions by one for each conversion.  I thought this was all fixed and working, but you've found a case not handled, or there has been some regression along the way.

You can try this yourself: take you sample input and paste it in some editor - Notepad++ works - and convert the EOL to Unix.  Then paste it back in the dnGrep test window sample text.  The highlights will be in the correct position with Multiline checked or not.

**Comment by fade2gray** [comment]
> dnGrep uses .NET regular expressions, and there is no g flag ...

... confusing as RegEx 101 allows toggling of the 'g' flag if the .Net Flavor is selected.

Also, I wonder if this is pertinent - [Stack Overflow](https://stackoverflow.com/questions/2715428/does-net-regex-support-global-matching#comment2741200_2715453)

**Comment by doug24** [comment]
I see Regex 101 is doing that, but I don't know how they are doing it, unless they are doing it in their code wrapper.  If I add a g option to a .NET regex, I get an invalid regular expression (both of these are invalid):
`(?g)bar`
`(?-g)bar` 
[.NET Reference](https://learn.microsoft.com/en-us/dotnet/standard/base-types/regular-expression-options)

dnGrep is already using Regex.Matches and iterating the groups to show the highlights.

**Comment by doug24** [comment]
Fixed in [4.1.47](https://github.com/dnGrep/dnGrep/releases/tag/v4.1.47.0)
